[BUG] drive list/info ignore pagination; info falsely reports File not found past page 1

## Summary
In a Feishu drive folder with more files than one API page, `feishu_drive` `action: info` returns a false `File not found: <token>` for any file that lives beyond the first page, and `action: list` cannot retrieve any page after the first. Both surfaces ignore the `has_more` / `next_page_token` cursor the Lark API returns, so files past page 1 are silently inaccessible.

## Environment
- Commit: cc451f98cbae5cb1db24cb3e97bced6d58952f57 (origin/main at time of filing)
- Surface: Feishu plugin `feishu_drive` tool, `list` and `info` actions (`extensions/feishu/src/drive.ts`, `extensions/feishu/src/drive-schema.ts`)

## Steps to reproduce
1. Have a Feishu drive folder with more files than fit in one `drive.file.list` page (the API signals this with `has_more: true` and a `next_page_token`).
2. Call `feishu_drive` `action: info` with the `file_token` of a file that sorts onto page 2 or later.
3. Or call `feishu_drive` `action: list` on that folder and try to fetch the next page.

## Expected
- `info` follows `has_more` / `next_page_token` until it finds the file, then returns its metadata.
- `list` accepts a page cursor so the caller can page through the whole folder, and reports whether more pages remain.

## Actual
- `info` searches only the first page and throws `File not found: <token>` for any file past page 1, even though the file exists and the API advertised more pages.
- `list` returns only the first page. The action schema exposes no `page_token`, so the returned `next_page_token` can never be fed back, and pages 2+ are unreachable.

## Root cause
`getFileInfo` (`extensions/feishu/src/drive.ts:356`) does a single first-page `.find()` and throws when absent, never following `has_more`:

```ts
async function getFileInfo(client: Lark.Client, fileToken: string, folderToken?: string) {
  // Use list with folder_token to find file info
  const res = await client.drive.file.list({
    params: folderToken ? { folder_token: folderToken } : {},
  });
  if (res.code !== 0) {
    throw new Error(res.msg);
  }

  const file = res.data?.files?.find((f) => f.token === fileToken);
  if (!file) {
    throw new Error(`File not found: ${fileToken}`);
  }
  // ...
}
```

`listFolder` (`extensions/feishu/src/drive.ts:331`) passes no `page_token` and the `list` action schema (`extensions/feishu/src/drive-schema.ts:25`) exposes none, so the `next_page_token` it does return at `drive.ts:352` can never be supplied on a follow-up call:

```ts
async function listFolder(client: Lark.Client, folderToken?: string) {
  const validFolderToken = folderToken && folderToken !== "0" ? folderToken : undefined;
  const res = await client.drive.file.list({
    params: validFolderToken ? { folder_token: validFolderToken } : {},
  });
  // ... returns next_page_token but the dispatch (drive.ts:770-773) accepts no cursor to send back
}
```

The Lark SDK `@larksuiteoapi/node-sdk` `drive.file.list` accepts `page_size` / `page_token` and returns `next_page_token` / `has_more` (and offers a `listWithIterator`). The plugin reads neither cursor field on this path.

## Sibling surfaces
This is the direct sibling-class of the just-merged wiki pagination fix #37626 (commit 5f90f08957), which paginated `wiki.spaceNode.list` / `wiki.space.list` in `wiki.ts` + `wiki-schema.ts` but did not touch `drive.ts`. The same plugin's comment paths already paginate correctly: `list_comments` and `list_comment_replies` expose `page_size` + `page_token` in the schema (`drive-schema.ts:54-66`) and read `has_more` / `page_token` from the response (`drive.ts:486-487`, `519-520`). So the drive `list` / `info` omission is the lone un-paginated lister in the plugin. Checked the open Feishu PRs (tool-family gates, card/proactive send, topic queue, audit redaction, URL challenge, streaming card coalescing); none touch drive `list` / `info`.

## Real behavior proof
Behavior addressed: `feishu_drive` `info` must find a file that lives past the first `drive.file.list` page instead of falsely reporting `File not found`.
Real environment tested: real `registerFeishuDriveTools` runtime driven end to end through the tool's `execute` dispatch and `getFileInfo`, with `client.drive.file.list` mocked to return the target on page 2 with `has_more: true` + `next_page_token`, at commit cc451f98cbae.
Exact steps or command run after this patch: `node scripts/run-vitest.mjs extensions/feishu/src/drive.pagination.repro.test.ts` (the repro registers the real drive tool, calls `action: info` for a token that sorts to page 2, and asserts the lister is drained to 2 pages and the file is returned).
Evidence after fix:
```
# OBSERVED (buggy, origin/main cc451f98cbae)
 FAIL  extensions/feishu/src/drive.pagination.repro.test.ts > info finds a file that lives past the first page
 AssertionError: expected "vi.fn()" to be called 2 times, but got 1 times
   115|     expect(listMock).toHaveBeenCalledTimes(2);
       |                      ^
 -> getFileInfo queried only page 1 (one list call) and threw "File not found: target";
    list was never asked for page 2 despite has_more: true.

# EXPECTED (correct behavior on identical inputs, with the local fix applied)
 Test Files  2 passed (2)
      Tests  16 passed (16)
 -> getFileInfo drains has_more/next_page_token to page 2 (two list calls), finds and
    returns the file. Existing drive.test.ts stays green (15 tests), 16 total.
```
Observed result after fix: on unmodified main, info silently fails for files past page 1 (list called once, throws File not found); draining pages following has_more/next_page_token fixes it without regressing existing drive behavior.
What was not tested: not run against a live Feishu tenant; the proof drives the real plugin code path with the Lark `drive.file.list` response shape mocked rather than a live multi-page folder. The `list` action was not exercised in the automated repro (its defect is structural: the schema exposes no `page_token` cursor, shown above), only `info` was driven end to end.

--- a/docs/channels/feishu.md
+++ b/docs/channels/feishu.md
@@ -324,6 +324,10 @@ The plugin ships agent tools for Feishu documents, chats, knowledge base, cloud 
 
 `tools.base` is an alias for `tools.bitable`; the explicit `bitable` value wins when both are set. Per-account gates live under `accounts.<id>.tools`.
 
+Grant `drive:drive.metadata:readonly` for direct `feishu_drive info` lookups outside the root
+directory, unless the app already has the full `drive:drive` scope. Without either scope, `info`
+keeps the legacy root-directory lookup available through `drive:drive:readonly`.
+
 ### ACP sessions
 
 Feishu/Lark supports ACP for DMs and group thread messages. Feishu/Lark ACP is text-command driven - there are no native slash-command menus, so use `/acp ...` messages directly in the conversation.

--- a/extensions/feishu/skills/feishu-drive/SKILL.md
+++ b/extensions/feishu/skills/feishu-drive/SKILL.md
@@ -48,7 +48,9 @@ root-list cursors are not forwarded.
 { "action": "info", "file_token": "ABC123", "type": "docx" }
 ```
 
-Searches for the file in the root directory. Note: file must be in root or use `list` to browse folders first.
+Looks up file metadata directly by token and type, regardless of which shared folder contains it.
+Shortcuts are the exception: Feishu's metadata API does not support the `shortcut` type, so shortcut
+info retains the root-directory lookup behavior.
 
 `type`: `doc`, `docx`, `sheet`, `bitable`, `folder`, `file`, `mindnote`, `shortcut`
 
@@ -101,7 +103,8 @@ channels:
 ## Permissions
 
 - `drive:drive` - Full access (create, move, delete)
-- `drive:drive:readonly` - Read only (list, info)
+- `drive:drive:readonly` - Read only (list and root-level info fallback)
+- `drive:drive.metadata:readonly` - Direct `info` lookup outside the root (not needed with `drive:drive`)
 
 ## Known Limitations
 

--- a/extensions/feishu/src/drive.test.ts
+++ b/extensions/feishu/src/drive.test.ts
@@ -436,6 +436,137 @@ describe("registerFeishuDriveTools", () => {
     });
   });
 
+  it("looks up file info directly by token and type", async () => {
+    const listFiles = vi.fn();
+    const batchQuery = vi.fn().mockResolvedValue({
+      code: 0,
+      data: {
+        metas: [
+          {
+            doc_token: "doc_1",
+            doc_type: "docx",
+            title: "Project Plan",
+            url: "https://example.test/doc_1",
+            create_time: "1710000000",
+            latest_modify_time: "1710001000",
+            owner_id: "ou_owner",
+          },
+        ],
+      },
+    });
+    createFeishuToolClientMock.mockReturnValue({
+      drive: { file: { list: listFiles }, meta: { batchQuery } },
+    });
+    const tool = buildDriveTool();
+
+    const result = await tool.execute("call-info", {
+      action: "info",
+      file_token: "doc_1",
+      type: "docx",
+    });
+
+    expect(batchQuery).toHaveBeenCalledWith({
+      data: {
+        request_docs: [{ doc_token: "doc_1", doc_type: "docx" }],
+        with_url: true,
+      },
+    });
+    expect(listFiles).not.toHaveBeenCalled();
+    expect(result.details).toEqual({
+      token: "doc_1",
+      name: "Project Plan",
+      type: "docx",
+      url: "https://example.test/doc_1",
+      created_time: "1710000000",
+      modified_time: "1710001000",
+      owner_id: "ou_owner",
+    });
+  });
+
+  it("reports a missing file when metadata lookup returns a failed entry", async () => {
+    const batchQuery = vi.fn().mockResolvedValue({
+      code: 0,
+      data: { metas: [], failed_list: [{ code: 970005, token: "missing_doc" }] },
+    });
+    createFeishuToolClientMock.mockReturnValue({
+      drive: { meta: { batchQuery } },
+    });
+    const tool = buildDriveTool();
+
+    const result = await tool.execute("call-info-missing", {
+      action: "info",
+      file_token: "missing_doc",
+      type: "docx",
+    });
+
+    expect(result.details).toMatchObject({ error: "File not found: missing_doc" });
+  });
+
+  it("keeps root-list lookup for shortcut info", async () => {
+    const batchQuery = vi.fn();
+    const listFiles = vi.fn().mockResolvedValue({
+      code: 0,
+      data: {
+        files: [
+          {
+            token: "shortcut_1",
+            name: "Project shortcut",
+            type: "shortcut",
+            url: "https://example.test/shortcut_1",
+          },
+        ],
+      },
+    });
+    createFeishuToolClientMock.mockReturnValue({
+      drive: { file: { list: listFiles }, meta: { batchQuery } },
+    });
+    const tool = buildDriveTool();
+
+    const result = await tool.execute("call-info-shortcut", {
+      action: "info",
+      file_token: "shortcut_1",
+      type: "shortcut",
+    });
+
+    expect(listFiles).toHaveBeenCalledWith({ params: {} });
+    expect(batchQuery).not.toHaveBeenCalled();
+    expect(result.details).toMatchObject({
+      token: "shortcut_1",
+      name: "Project shortcut",
+      type: "shortcut",
+    });
+  });
+
+  it("falls back to root lookup when the metadata scope is missing", async () => {
+    const batchQuery = vi.fn().mockRejectedValue({
+      response: {
+        data: {
+          code: 99991672,
+          msg: "permission denied: drive:drive.metadata:readonly",
+        },
+      },
+    });
+    const listFiles = vi.fn().mockResolvedValue({
+      code: 0,
+      data: {
+        files: [{ token: "doc_1", name: "Project Plan", type: "docx" }],
+      },
+    });
+    createFeishuToolClientMock.mockReturnValue({
+      drive: { file: { list: listFiles }, meta: { batchQuery } },
+    });
+    const tool = buildDriveTool();
+
+    const result = await tool.execute("call-info-missing-scope", {
+      action: "info",
+      file_token: "doc_1",
+      type: "docx",
+    });
+
+    expect(listFiles).toHaveBeenCalledWith({ params: {} });
+    expect(result.details).toMatchObject({ token: "doc_1", name: "Project Plan", type: "docx" });
+  });
+
   it("normalizes folder pagination and suppresses it for root listings", async () => {
     const listFiles = vi.fn().mockResolvedValue({ code: 0, data: { files: [] } });
     createFeishuToolClientMock.mockReturnValue({ drive: { file: { list: listFiles } } });

--- a/extensions/feishu/src/drive.ts
+++ b/extensions/feishu/src/drive.ts
@@ -367,16 +367,13 @@ async function listFolder(client: Lark.Client, params: Record<string, unknown> =
   };
 }
 
-async function getFileInfo(client: Lark.Client, fileToken: string, folderToken?: string) {
-  // Use list with folder_token to find file info
-  const res = await client.drive.file.list({
-    params: folderToken ? { folder_token: folderToken } : {},
-  });
+async function getRootFileInfo(client: Lark.Client, fileToken: string) {
+  const res = await client.drive.file.list({ params: {} });
   if (res.code !== 0) {
     throw new Error(res.msg);
   }
 
-  const file = res.data?.files?.find((f) => f.token === fileToken);
+  const file = res.data?.files?.find((candidate) => candidate.token === fileToken);
   if (!file) {
     throw new Error(`File not found: ${fileToken}`);
   }
@@ -388,6 +385,58 @@ async function getFileInfo(client: Lark.Client, fileToken: string, folderToken?:
     url: file.url,
     created_time: file.created_time,
     modified_time: file.modified_time,
+    owner_id: file.owner_id,
+  };
+}
+
+async function getFileInfo(
+  client: Lark.Client,
+  fileToken: string,
+  type: Extract<FeishuDriveParams, { action: "info" }>["type"],
+) {
+  if (type === "shortcut") {
+    // The metadata API does not accept shortcut as a document type. Keep the existing
+    // root-list behavior so the advertised shortcut info contract does not regress.
+    return getRootFileInfo(client, fileToken);
+  }
+
+  let res: Awaited<ReturnType<Lark.Client["drive"]["meta"]["batchQuery"]>>;
+  try {
+    res = await client.drive.meta.batchQuery({
+      data: {
+        request_docs: [{ doc_token: fileToken, doc_type: type }],
+        with_url: true,
+      },
+    });
+  } catch (error) {
+    if (extractDriveApiErrorMeta(error).feishuCode === 99991672) {
+      // Existing read-only apps may not have the newer metadata scope. Preserve their
+      // root-file lookup while allowing scoped apps to resolve files in any shared folder.
+      return getRootFileInfo(client, fileToken);
+    }
+    throw error;
+  }
+  if (res.code === 99991672) {
+    return getRootFileInfo(client, fileToken);
+  }
+  if (res.code !== 0) {
+    throw new Error(res.msg);
+  }
+
+  const file = res.data?.metas?.find(
+    (meta) => meta.doc_token === fileToken || meta.request_doc_info?.doc_token === fileToken,
+  );
+  if (!file) {
+    throw new Error(`File not found: ${fileToken}`);
+  }
+
+  return {
+    token: file.doc_token,
+    name: file.title,
+    type: file.doc_type,
+    url: file.url,
+    created_time: file.create_time,
+    modified_time: file.latest_modify_time,
     owner_id: file.owner_id,
   };
 }
@@ -791,7 +840,7 @@ export function registerFeishuDriveTools(api: OpenClawPluginApi) {
                   }),
                 );
               case "info":
-                return jsonResult(await getFileInfo(client, p.file_token));
+                return jsonResult(await getFileInfo(client, p.file_token, p.type));
               case "create_folder":
                 return jsonResult(await createFolder(client, p.name, p.folder_token));
               case "move":

